### PR TITLE
QDOC: links

### PR DESCRIPTION
**Pull request by Undin**
Support links in quick documentation.

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/docs/DocumentationUtils.kt [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+package org.rust.ide.docs
+
+import com.intellij.codeInsight.documentation.DocumentationManagerUtil
+import com.intellij.psi.PsiElement
+import org.rust.ide.utils.escaped
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+
+internal fun PsiElement.generateDocumentation(buffer: StringBuilder, prefix: String = "", suffix: String = "") {
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/docs/DocumentationUtils.kt]: `internal` doesn't make sense in our codebase, as it means *private outside module* not *private outside package* as Java's `package-local` does

Let's just make it public, eventually leave a comment that this function is package internal

File: src/main/kotlin/org/rust/ide/docs/DocumentationUtils.kt [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            buffer += "&gt;::"
+        }
+        else -> buffer += text.escaped
+    }
+    buffer += suffix
+}
+
+private fun generatePathDocumentation(element: RsPath, buffer: StringBuilder) {
+    val path = element.path
+    path?.let {
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/docs/DocumentationUtils.kt]: nitpicking: `if(path != null)` maybe? seems slightly more readable to me

File: src/main/kotlin/org/rust/ide/docs/DocumentationUtils.kt [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+private fun RsPath.isLinkNeeded(): Boolean {
+    val element = reference.resolve()
+    // If it'll find out that links for type parameters are useful
+    // just check element for null
+    return !(element == null || element is RsTypeParameter)
+}
+
+private fun RsPath.link(): String {
+    val path = path
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/docs/DocumentationUtils.kt]: ❤️

File: src/main/kotlin/org/rust/ide/docs/DocumentationUtils.kt (hunk 1 above)
Comment (Undin) [inline src/main/kotlin/org/rust/ide/docs/DocumentationUtils.kt]: Hm, indeed.
Maybe drop `DocumentataionUtils` at all and move all functions into `RsDocumentationProvider` as private functions? It'll make `RsDocumentationProvider.kt` bigger but allow not to provide internal methods into global namespace.

**Comment by Undin** [comment]
Comments are fixed

File: src/main/kotlin/org/rust/ide/docs/DocumentationUtils.kt (hunk 1 above)
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/docs/DocumentationUtils.kt]: seems fine to me

File: src/main/kotlin/org/rust/ide/docs/RsDocumentationProvider.kt [hunk 4]
```diff
@@ -45,10 +47,22 @@ class RsDocumentationProvider : AbstractDocumentationProvider() {
 
     private fun generateDoc(element: RsTypeParameter): String? {
         val name = element.name ?: return null
+        val buffer = StringBuilder("type parameter ")
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/docs/RsDocumentationProvider.kt]: I wish there were extension `+=` for `StringBuilder` in Kotlin like in Scala

**Comment by matklad** [comment]
Hm, overall, I am not quite happy with the logic in `PresentaionUtil.kt` and `RsDocumentationProvider`: there are just too many casts, and everything looks extremely brittle. I don't have a good plan of how to fix this, except for writing more tests though :)

**Comment by matklad** [comment]
Oups, sorry, I've pushed some refactorings in the documentation area and I've totally forgot that it'll conflict with this PR :(

**Comment by Undin** [comment]
No problem:)
I'll just rebase

**Comment by Undin** [comment]
@matklad Do you have cases what you want to check in tests? If yes I'll be glad to add them.
At this moment there are more than 100 tests for `RsDocumentationProvider` and in my opinion, they test all common cases.

**Comment by matklad** [comment]
@Undin no, I think the current tests cover most of the reasonable use cases. The problem is, it's difficult to make sure that weird use-cases also work. Recently, after the `navigationElement` fix, similar code in Presentation utils start throwing assertion errors when opening a real project, despite the green tests: https://github.com/intellij-rust/intellij-rust/commit/9a93b97e9c8ac7ff4409c31495786c438739a854

I don't really know how we can improve situation here though, other then carefully scrutinize our PSI hierrarchy and try to extract some more focus intefaces to cut down on `when (randomPsiElement)` switches.

**Pull request merged**